File: backend/app/schemas/transport_excise.py
"""Wire schemas for the diesel excise-duty refund (WO-91, G4.6 — R42, R53).

Field-for-field from `app/services/transport/excise.py` — this module shapes the
wire and computes nothing (engineering-rules §3). Every rate and every euro is
typed `Decimal`, so pydantic v2 serializes it as a JSON STRING and no float ever
appears in a money path (§4.9). `litres` is a `Decimal` too: it is the figure's
multiplicand and a float round-trip of a 3-decimal litre total would move the
euro computed from it.

THE ELIGIBILITY LIMITATION RIDES THE WIRE, IT IS NOT A UI DECORATION
-----------------------------------------------------------------------
R42's acceptance line is *"The UI shows the indicative-rate and eligibility
caveats on **every surface that shows the number**"*, and `BA_fleet_fuel.md`
§3.L is the rule behind it: *"`excise.py` | **Asserts NO eligibility** (vehicle
>=7.5t / carrier registration not modelled); rates are indicative defaults."*

So `eligibility`, `rate_caveat`, `legal_framing` and `eligibility_asserted` are
**required** fields on every response that carries a euro. They are not optional
metadata a client may drop: a surface physically cannot render the number
without having received the denial next to it, and `eligibility_asserted` is a
literal `False` so the denial survives a UI that truncates prose. The strings
come from `excise.ELIGIBILITY_STATEMENT` / `excise.RATE_CAVEAT` /
`excise.LEGAL_FRAMING` — one definition each, in the service, so no wording can
be softened downstream.

**No field name in this module contains `recover`, `owed`, `owes`, `claim`,
`demand`, `due`, `debt` or `payable`**, and
`tests/transport/test_wo91_excise_eligibility.py` asserts it against the SAME
word list WO-87 wrote for the overpay surface. That is not a style rule: a
`recoverable_eur` field here would read as an entitlement the moment a client
rendered it, and whether this haulier is entitled to anything is precisely what
this product does not model. The euro is `indicative_excise_eur` — the
qualification lives inside the field name, where a renderer cannot lose it.
"""

from __future__ import annotations

from decimal import Decimal

from pydantic import BaseModel, Field


class ExciseRateOut(BaseModel):
    """One state's rate as the figure will actually use it: the org's typed
    override where there is one, the harvested indicative default otherwise.

    `is_override` is what distinguishes *"an operator verified this with
    customs"* from *"this is the spec's EUR 30.00 placeholder"* — the whole
    reason §9.2 item 13 exists as an open question. A surface that shows the
    rate without it would be presenting a placeholder as a statutory figure.
    """

    country: str
    rate_eur_per_1000l: Decimal
    is_override: bool
    rate_caveat: str


class ExciseRatesOut(BaseModel):
    """Every state this product records as operating the regime, with its
    resolved rate. Carries the caveats so the rate table is itself one of R42's
    *"surfaces that show the number"*."""

    countries: list[str]
    default_rate_eur_per_1000l: Decimal
    rates: list[ExciseRateOut]
    eligibility: str
    eligibility_asserted: bool
    rate_caveat: str
    legal_framing: str
    currency: str


class ExciseCellOut(BaseModel):
    """One (entity × country) cell — R42's grain. `litres` is a `Decimal`
    because it is the figure's multiplicand: a float round-trip of a
    three-decimal litre total would move the euro computed from it.

    `rate_is_override` distinguishes a rate a human verified with customs from
    the harvested EUR 30.00 placeholder. A packet that could not tell them apart
    would present a placeholder as a statutory figure.
    """

    entity_id: str
    entity_name: str
    country: str
    litres: Decimal
    rate_eur_per_1000l: Decimal
    rate_is_override: bool
    indicative_excise_eur: Decimal
    lines: int


class SkippedCountryOut(BaseModel):
    """A country with validated diesel litres in scope for which this product
    holds no rate. It carries litres and a line count and **no euro** — there is
    no figure, which is not the same as a figure of zero."""

    country: str
    litres: Decimal
    lines: int


class ExciseReportOut(BaseModel):
    """The diesel excise-duty refund over one accounting month.

    Every caveat field is REQUIRED — see the module docstring. `filed_with`
    names the addressee because §2.4/§1.2/§6.1 all stress that this is a
    SEPARATE regime from the VAT refund, claimed from customs; a surface that
    lost that would invite the figure onto a VAT filing.
    """

    period: str
    entity_id: str | None
    country: str | None
    currency: str
    product_group: str
    litre_basis: str
    legal_framing: str
    eligibility: str
    eligibility_asserted: bool
    rate_caveat: str
    filed_with: str
    rows: list[ExciseCellOut]
    skipped_countries: list[SkippedCountryOut]
    litres: Decimal
    indicative_excise_eur: Decimal
    lines_examined: int


class ExciseRateIn(BaseModel):
    """Type the rate a customs authority actually applies for one state.

    The country and the positive-rate rule are enforced in the SERVICE
    (`excise_country_not_supported` / `invalid_excise_rate`) as well as bounded
    here — schemas catch shape, services catch business rules (master-context
    DoD 3). `gt=0` is the shape half; the service's refusal explains WHY a zero
    is not "this state refunds nothing".
    """

    country: str = Field(min_length=2, max_length=2)
    rate_eur_per_1000l: Decimal = Field(gt=0)

File: frontend/src/lib/transportExcise.ts
/**
 * Pure helpers for the diesel-excise screen (WO-92) over the five WO-91 routes.
 * No React, no network, no arithmetic — so the vocabulary this screen renders is
 * testable through the page and cannot drift into a component.
 *
 * WHAT THIS FILE IS REALLY FOR: THE ELIGIBILITY NON-ASSERTION
 * ------------------------------------------------------------
 * Around seven EU states refund part of the excise a haulier pays on commercial
 * diesel. This product multiplies validated diesel litres by a per-1,000-litre
 * rate and stops there. The conditions that decide whether a given haulier
 * actually qualifies — vehicle weight and carrier registration — are
 * deliberately NOT modelled, and `app/services/transport/excise.py` says so in
 * one server-side sentence that rides every response.
 *
 * Which is why this module holds NO framing string of its own. The eligibility
 * statement, the rate caveat, the legal framing, the customs addressee and the
 * litre basis are all rendered VERBATIM from the wire, exactly as
 * `transportSavings.ts` deliberately holds none of its own: a paraphrase is how
 * a limitation gets softened, and there is intentionally only one string. What
 * lives here instead is the surrounding explanation — why a country we hold no
 * rate for produces nothing rather than a zero, what changing a rate does, and
 * what the screen does to the underlying data (nothing) — none of which restates
 * the server's wording and none of which borrows a word from the family of
 * vocabulary this surface must never use.
 *
 * THE RATE IS A PLACEHOLDER UNTIL SOMEONE VERIFIES IT. The shipped default is an
 * explicit placeholder in a reported band, not a statutory figure: the duty is
 * not harmonised across the EU and national rates are revised. So every rate the
 * screen shows is shown beside its SOURCE, and `RATE_SOURCE_LABELS` is that pair
 * of words — the same two the customs workbook prints, so the spreadsheet and
 * the screen say the same thing about the same rate.
 *
 * NOTHING HERE TOUCHES A NUMBER. Every figure the screen shows arrives already
 * multiplied and already quantized by the service and is rendered from its wire
 * string (§4.9/§4.10). There is no arithmetic of any kind in this module.
 */

/** The two panels, in the order the service's own module presents them: the
 * period's figures first, the rate registry they are computed from second. */
export const EXCISE_TABS: { value: ExciseTabKey; label: string }[] = [
  { value: "figures", label: "Monthly figures" },
  { value: "rates", label: "Country rates" },
];

export type ExciseTabKey = "figures" | "rates";

/**
 * The two words a rate's source may read, keyed by the wire's own boolean
 * (`rate_is_override` on a cell, `is_override` on a rate row). They are the
 * customs workbook's own pair, so the sheet a haulier files and the screen an
 * operator reads describe the same rate identically.
 */
export const RATE_SOURCE_LABELS = {
  override: "Verified override",
  placeholder: "Indicative default",
} as const;

export function rateSourceLabel(isOverride: boolean): string {
  return isOverride
    ? RATE_SOURCE_LABELS.override
    : RATE_SOURCE_LABELS.placeholder;
}

/**
 * The presentation rule this surface is most likely to get wrong, and the one
 * the whole screen is arranged around.
 *
 * A country this product holds no rate for produces NO figure. Printing €0.00
 * for it would say "this state refunds you nothing", which is a different
 * statement from "we hold no rate for this state" and is materially false. The
 * wire shape for those countries carries litres and a line count and no euro
 * field at all, so the table below cannot render one even by accident.
 */
export const NO_RATE_EXPLANATION =
  "These countries had validated diesel litres in the month, and this product holds no rate for them. " +
  "They are listed with their litres so the size of what is not being reported is visible. " +
  "That is not a figure of zero: zero would mean the state refunds nothing, and what is actually true is " +
  "that we hold no rate to compute anything with. Type a rate under Country rates and they will appear above.";

/**
 * The rate half of the same honesty, in the operator's words rather than the
 * service's (the service's own `rate_caveat` is rendered verbatim beside this).
 * Stated because a rate shown without its provenance reads as a statutory
 * figure, and the shipped one is not.
 */
export const RATE_SOURCE_NOTE =
  "Every rate on this screen is shown with its source. An indicative default is the placeholder this product ships " +
  "for a state nobody has verified yet; a verified override is a rate an operator typed after checking it with the " +
  "customs authority. Overriding a rate changes only what the indicative figure is computed from — it files nothing, " +
  "sends nothing and changes no other figure in the workspace.";

/** What this screen does to the underlying data: nothing (§4.19). The rate
 * registry is the one exception and it is configuration, not a financial
 * record. */
export const ADVISORY_NOTE =
  "The analysis and the customs packet are read-only. Opening this page changes no figure, no status and no document — " +
  "it re-reads the validated diesel lines, sums their litres and applies the rate, and nothing here gates or blocks any " +
  "other part of the workspace. Setting or clearing a country rate is the only change this screen can make, and it is " +
  "recorded against the operator who made it.";

/** What the packet is and who it goes to — the addressee itself comes off the
 * wire (`filed_with`); this is the surrounding explanation of the artifact. */
export const PACKET_NOTE =
  "The evidence packet is the spreadsheet a haulier files with customs: the same litres, the same rate and the same " +
  "figure as the table above, with a total row and the same caveats printed on both sheets. It is generated from this " +
  "month's analysis, so it cannot disagree with what is on screen. A month with nothing to report produces no packet — " +
  "an empty filing looks like a filing and supports nothing.";
